Feature/flow 579 add stateid into dummy service so that CRUD operations only affects data in that state."

xiaoqianshi-dell opened the pull request:
Adding stateid into account type as a property so that when we do CRUD operations, sql queries will only be applied to the data created in that state. So I am no longer adding stateid into value name as originally planned, hopefully this is a better way.

xiaoqianshi-dell committed:
recreate table again

xiaoqianshi-dell committed:
Revert "recreate table again"

This reverts commit d56cc97e

xiaoqianshi-dell committed:
feature/FLOW-579 add stateid into dummy service so that CRUD operations only affects data in that state.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -16,8 +16,8 @@ public AccountRepository(Sql2o sql2o) {
-    public Account find(String id) {
-        String sql = "SELECT * FROM account WHERE id = :id";
+    public Account find(String id ) {
+        String sql = "SELECT * FROM account WHERE id = :id where stateid = :stateid";
```
There are two "where" in this query, also because the id is unique you don't need the satateid part.

xiaoqianshi-dell replied:
Removed stateid as suggested.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -36,26 +36,27 @@ public Account find(String id) {
 
     public Account update(Account account) {
-        String sql = "UPDATE account SET name = :name, company = :company WHERE id = :id";
+        String sql = "UPDATE account SET name = :name, company = :company, stateid = :stateid WHERE id = :id and stateid = :stateid ";
```
the same here, the id should be enough to identify the row that you are modifying

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -65,8 +66,8 @@ public Account create(Account account) {
-    public void delete(String id) {
-        String sql = "DELETE FROM account WHERE id = :id";
+    public void delete(String id ) {
+        String sql = "DELETE FROM account WHERE id = :id and stateid = :stateid";
```
The same here

xiaoqianshi-dell replied:
Removed stateid as suggested.

josecollazzi reviewed (changes requested):
I have request some changes in the queries.

xiaoqianshi-dell committed:
feature/FLOW-579 remove unnecessary stateid from WHERE clause.

xiaoqianshi-dell committed:
feature/FLOW-579 adding filter  CriteriaType to Account data set.

xiaoqianshi-dell committed:
feature/FLOW-579 move else statement.

xiaoqianshi-dell committed:
feature/FLOW-579 modified filtering again and added exception handling to throw unsupported error message to engine

xiaoqianshi-dell commented:
I've modified findAll method the AccountRepository class for filtering table columns / different object properties. Also added a little bit error handling.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,104 @@ public AccountRepository(Sql2o sql2o) {
-    public List<Account> findAll(ListFilter filter) {
-        String sql = "SELECT * FROM account";
-
+    public Account findAll(String filterBy, String value) {
```
this method is not used, you can get rid of it

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,104 @@ public AccountRepository(Sql2o sql2o) {
+
+    public List<Account> findAll(ListFilter filter) throws ServiceProblemException {
+
+        String sql = "";
```
all the queries start by `SELECT * FROM account` could you initialise to that value, I think it would look more simple.

xiaoqianshi-dell replied:
Yes, I am now initialising this value with default query.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,104 @@ public AccountRepository(Sql2o sql2o) {
+
+        String sql = "";
+        if (filter != null && filter.getWhere() != null && filter.getWhere().size() != 0) {
+            String columnname = filter.getWhere().get(0).getColumnName();
```
could you change to camel case here, this code only works when filtering with one value I think it could be easy to allow filter by many values. But if you want to keep it very simple you should throw an error if the size is > 1

xiaoqianshi-dell replied:
I've added another if condition to throw an error when size > 1.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,104 @@ public AccountRepository(Sql2o sql2o) {
+
+                if (criteriaType == CriteriaType.GreaterThan) {
+                    sql = "SELECT * FROM account WHERE id > " + value;
+                } else if (criteriaType == CriteriaType.GreaterThanOrEqual) {
+                    sql = "SELECT * FROM account WHERE id >= " + value;
+                } else if (criteriaType == CriteriaType.LessThan) {
+                    sql = "SELECT * FROM account WHERE id < " + value;
+                } else if (criteriaType == CriteriaType.LessThanOrEqual) {
+                    sql = "SELECT * FROM account WHERE id <= " + value;
+                } else if (criteriaType == CriteriaType.Equal) {
+                    sql = "SELECT * FROM account WHERE id == " + value;
+                } else if (criteriaType == CriteriaType.NotEqual) {
+                    sql = "SELECT * FROM account WHERE id != " + value;
+                } else if (criteriaType == CriteriaType.IsEmpty) {
+                    sql = (Boolean.valueOf(filter.getWhere().get(0).getContentValue())) ?
+                            "SELECT * FROM account WHERE id is null" :
+                            "SELECT * FROM account WHERE id is not null";
+                } else {
+                    throw new ServiceProblemException(400, "Integer type \"id\" only supports EQUAL, NOT_EQUAL, GREATER_THAN, GREATER_THAN_OR_EQUAL, LESS_THAN, LESS_THAN_OR_EQUAL and IS_EMPTY operations");
+                }
+            } else if (columnname.equals("company") || columnname.equals("name") || columnname.equals("stateid")) {
+
+                if (criteriaType == CriteriaType.StartsWith) {
+                    sql = "SELECT * FROM account WHERE " + columnname + " LIKE \'" + value + "%\'";
+                } else if (criteriaType == CriteriaType.EndsWith) {
+                    sql = "SELECT * FROM account WHERE " + columnname + " LIKE \'%" + value + "\'";
+                } else if (criteriaType == CriteriaType.IsEmpty) {
+                    sql = (Boolean.valueOf(filter.getWhere().get(0).getContentValue())) ?
+                            "SELECT * FROM account WHERE " + columnname + " is null" :
+                            "SELECT * FROM account WHERE " + columnname + " is not null";
+                } else if (criteriaType == CriteriaType.Contains) {
+                    sql = "SELECT * FROM account WHERE " + columnname + " LIKE \'%" + value + "%\'";
+                } else if (criteriaType == CriteriaType.Equal) {
+                    sql = "SELECT * FROM account WHERE " + columnname + " == \'" + value + "\'";
+                } else if (criteriaType == CriteriaType.NotEqual) {
+                    sql = "SELECT * FROM account WHERE " + columnname + " != \'" + value + "\'";
+                } else {
+                    throw new ServiceProblemException(400, "String CriteriaType only supports EQUAL, NOT_EQUAL, STARTS_WITH, ENDS_WITH, CONTAINS and IS_EMPTY operations.");
+                }
+            }
```
This part of the code is not really reusable, I think it would be better if you create the query in one method and validate the criteria in other. Also you can use a switch that is more easy to read.

xiaoqianshi-dell replied:
I've extracted this block code into two methods , one for integer, one for string type, is this the right way to do it?
Also change from if to switch to make it easier to read.

josecollazzi replied:
yes that will work too.

josecollazzi reviewed (changes requested):
I have suggested some changes

xiaoqianshi-dell committed:
feature/FLOW-579 extracted sql creation to a separate method and removed unused findAll method

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,115 @@ public AccountRepository(Sql2o sql2o) {
+        if (columnName.equals("id")) {
+            switch (criteriaType) {
+                case GreaterThan:
+                    sql = sql + " WHERE " + columnName + " > " + value;
```
the `where` word is added in each case except if there is an exception, you can added before the switch, and do the same in the other switch, also it looks to me a good case to use StringBuilder and pass the StringBuilder instead of a String
It can be used like this:
```       
 StringBuilder stringBuilder = new StringBuilder("SELECT * FROM account");
        stringBuilder.append(" WHERE ");
        stringBuilder.append("a = b");
        stringBuilder.toString();
```

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,115 @@ public AccountRepository(Sql2o sql2o) {
+            sql = createSQL(filter, sql, columnName, filterByValue, criteriaType);
+        }
+        try (
+                Connection connection = sql2o.open()) {
```
this line tab is not aligned

josecollazzi reviewed (changes requested):
Nice, just a minor changes.

xiaoqianshi-dell committed:
feature/FLOW-579 refactoring the code a bit by using StringBuilder and making createSQL method handling all cases.

xiaoqianshi-dell commented:
Thanks for suggesting to use StringBuilder, it indeed makes code look neater. I've also moved a few condition comparisons into createSQL method to make stuff more encapsulated.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,144 @@ public AccountRepository(Sql2o sql2o) {
+                        stringBuilder.append(value);
+                        break;
+                    case IsEmpty:
+                        stringBuilder.append(" > ");
```
this line shouldn't be there

xiaoqianshi-dell replied:
Removed these two lines as suggested.

josecollazzi commented on src/main/java/com/manywho/services/dummy/account/repositories/AccountRepository.java:
```diff
@@ -18,44 +20,144 @@ public AccountRepository(Sql2o sql2o) {
+                        stringBuilder.append(" LIKE \'");
+                        stringBuilder.append(value);
+                        stringBuilder.append("%\'");
+                        break;
```
The three lines above can be replaced by String format, that is usually more easy to read:

````
stringBuilder.append(String.format(" LIKE '%s%%'", value);
````

Or you can also append in each case the column instead of line 53, adding the value so it will be more easy to follow
 e.g.

````
stringBuilder.append(String.format("%s LIKE '%s%%' ", columnName, value);
````

This advice is valid for all the case blocks.

xiaoqianshi-dell replied:
Thanks Jose, String.format indeed makes the code look a lot better. And I've moved columnName into String.format expression to make it more readable.

xiaoqianshi-dell committed:
feature/FLOW-579 use String.format to simplify stringBuilder code.

josecollazzi reviewed (approved):
Nice work.
